addons/website/models/website.py

```diff
@@ -2076,68 +2118,38 @@ def _trigram_enumerate_words(self, search_details, search, limit):
-                ))
-            similarities.extend(indirect_similarities)
-            best_similarity = SQL('GREATEST(%(similarities)s)', similarities=SQL(', ').join(similarities))
-
-            # Filter unpublished records for portal and public user for
-            # performance.
-            # TODO: Same for `active` field?
-            filter_is_published = (
-                'is_published' in model._fields
-                and model._fields['is_published'].base_field.model_name == model_name
-                and not self.env.user._is_internal()
-            )
-            if filter_is_published:
-                query.add_where('is_published')
-
-            query.order = '_best_similarity desc'
-            query.limit = 1000
-            self.env.cr.execute(query.select(
-                SQL.identifier(model._table, 'id'),
-                SQL('%s AS _best_similarity', best_similarity),
-            ))
-            ids = {row[0] for row in self.env.cr.fetchall() if row[1] and row[1] >= similarity_threshold}
+            # Group indirect_fields by comodel
+            indirect_fields_info = defaultdict(dict)
+            for name, indirect_field in indirect_fields.items():
+                indirect_fields_info[indirect_field['comodel']][name] = indirect_field
+            subqueries = []
+            subqueries.append(get_subquery(model, direct_fields, 'id'))
+            for comodel in indirect_fields_info:
+                comodel_similarity_fields = set()
+                id_column = ''
+                rel_table = ''
+                rel_joinkey = ''
+                for indirect_field_info in indirect_fields_info[comodel].values():
+                    direct_field = model._fields[indirect_field_info['direct']]
+                    if direct_field.type == 'one2many':
+                        comodel_similarity_fields.add(indirect_field_info['indirect'])
+                        id_column = indirect_field_info['cofield']
+                    elif direct_field.type == 'many2many':
+                        comodel_similarity_fields.add(indirect_field_info['indirect'])
+                        id_column = direct_field.column1
+                        rel_table = direct_field.relation
+                        rel_joinkey = direct_field.column2
+                subqueries.append(get_subquery(comodel, comodel_similarity_fields, id_column, rel_table, rel_joinkey))
+            query = SQL("""
+                SELECT id,
+                    MAX(similarity) as _best_similarity
+                FROM (%s)
+                GROUP BY id
+                ORDER BY _best_similarity DESC
+                LIMIT 1000
+            """, SQL("\nUNION ALL\n").join(subqueries))  # UNION ALL allows to hit GIST indexes in subplans.
+            self.env.cr.execute(query)
+            ids = {row[0] for row in self.env.cr.fetchall()}
             domain.append([('id', 'in', list(ids))])
             domain = AND(domain)
```

Comment:
This is unrelated to your PR, as it's like that before the diff, but there might be a bug? `domain` is defined to always be the last domain copied from the main loop of the `search_details` (and if search_details is empty, it's a crash). Just taking a note, not sure what was the intention originally.

Reply:
I'm not sure we should fixed this here since it's outside of this PR's diff

Reply:
yeah it was just an observation